sdmetrics/single_table/privacy/dcr_overfitting_protection.py

```
@@ -0,0 +1,177 @@
+                A pandas.DataFrame object containing the synthetic data sampled
+                from the synthesizer.
+            real_validation_data (pd.DataFrame):
+                A pandas.DataFrame object containing a validation set of real data.
+                This data should not have been used to train the synthesizer.
+            metadata (dict):
+                A metadata dictionary that describes the table of data.
+            num_rows_subsample (int or None):
+                The number of synthetic data rows to subsample from the synthetic data.
+                This is used to increase the speed of the computation, if the dataset is large.
+                Defaults to None which means no subsampling will be done.
+            num_iterations (int):
+                The number of iterations to perform when subsampling.
+                The final score will be the average of all iterations. Default is 1 iteration.
+
+        Returns:
+            dict:
+                Returns a dictionary that contains the overall score, the % of synthetic data rows
+                that were closer to the validation set, and the % of synthetic data rows that were
+                closer to the real dataset. Averages of the medians are returned in the case of
+                multiple iterations.
+        """
+        sanitized_data = cls._validate_inputs(
+            real_training_data,
+            synthetic_data,
+            real_validation_data,
+            metadata,
+            num_rows_subsample,
+            num_iterations,
+        )
+
+        training_data, sanitized_synthetic_data, validation_data = sanitized_data
+
+        sum_of_scores = 0
+        sum_of_p_close_to_real = 0
+        sum_of_p_close_to_random = 0
+        for _ in range(num_iterations):
+            synthetic_sample = sanitized_synthetic_data
+            if num_rows_subsample is not None:
+                synthetic_sample = sanitized_synthetic_data.sample(n=num_rows_subsample)
```

What happens if `num_rows_subsample` is larger than our synthetic data?

---

I'll add a check for that too so it fails fast.

---

Instead of erroring out can we warn the user that the data is smaller than the desired subsample and just continue without subsampling?

---

Yep that makes sense! I added tests for that case too.